### PR TITLE
CUAEV double backward for force training

### PR DESCRIPTION
**Pull request by yueyericardo**

### CODE REVIEW

**Comment by yueyericardo** [comment]
benchmark of ani1x on gtx1080
batchsize:1500

```
RUN                        Total AEV    Forward      Backward     Force        Optimizer    Others       Epoch time   GPU
0 cu Energy train          8.6 ms       17.1 ms      17.2 ms      0.0 ms       21.3 ms      4.7 ms       68.9 ms      881.5MB
1 py Energy train          25.5 ms      13.8 ms      18.0 ms      0.0 ms       16.2 ms      4.3 ms       77.8 ms      975.5MB
2 cu Energy + Force train  7.4 ms       13.8 ms      57.5 ms      33.3 ms      16.7 ms      6.8 ms       135.6 ms     1085.5MB
3 py Energy + Force train  22.5 ms      11.6 ms      124.2 ms     61.7 ms      20.1 ms      8.3 ms       248.4 ms     1425.5MB
```

File: tests/test_cuaev.py [hunk 1]
```diff
@@ -89,15 +91,121 @@ def testSimpleBackward(self):
...
+        species = torch.tensor([[1, 0, 0, 0, 0], [2, 0, 0, 0, -1]], device=self.device)
+
+        self.nn.zero_grad()
+        _, aev = self.aev_computer((species, coordinates))
+        E = self.nn(aev).sum()
+        force_ref = -torch.autograd.grad(E, coordinates, create_graph=True, retain_graph=True)[0]
+        force_true = torch.randn_like(force_ref)
+        loss = torch.abs(force_true - force_ref).sum(dim=(1, 2)).mean()
+        loss.backward()
+        param = next(self.nn.parameters())
```
Comment (zasdfgbnm) [inline tests/test_cuaev.py:129]: Instead of copy-pasting this part twice, could you make it a function and call it?

File: tests/test_cuaev.py [hunk 2]
```diff
@@ -89,15 +91,121 @@ def testSimpleBackward(self):
...
+        aev_grad = torch.autograd.grad(E, aev_, create_graph=True, retain_graph=True)[0]
+        # graph1 backward
+        aev_grad_ = aev_grad.clone().detach().requires_grad_()
+        force_ref = torch.autograd.grad(aev, coordinates, aev_grad_, create_graph=True, retain_graph=True)[0]
+        # force loss backward
+        force_true = torch.randn_like(force_ref)
+        loss = torch.abs(force_true - force_ref).sum(dim=(1, 2)).mean()
+        aev_grad_grad = torch.autograd.grad(loss, aev_grad_, create_graph=True, retain_graph=True)[0]
+        aev_grad_grad_radial = aev_grad_grad[0][0][:self.radial_length].view(-1, self.aev_computer.radial_sublength)
+        aev_grad_grad_angular = aev_grad_grad[0][0][self.radial_length:].view(-1, self.aev_computer.angular_sublength)
```
Comment (zasdfgbnm) [inline tests/test_cuaev.py:184]: Same as above, make it a function and call it twice

File: tests/test_cuaev.py [hunk 3]
```diff
@@ -129,6 +237,47 @@ def testTripeptideMDBackward(self):
...
+                E = self.nn(aev_).sum()
+                # graph2 backward
+                aev_grad = torch.autograd.grad(E, aev_, create_graph=True, retain_graph=True)[0]
+                # graph1 backward
+                aev_grad_ = aev_grad.clone().detach().requires_grad_()
+                force_ref = torch.autograd.grad(aev, coordinates, aev_grad_, create_graph=True, retain_graph=True)[0]
+                # force loss backward
+                force_true = torch.randn_like(force_ref)
+                loss = torch.abs(force_true - force_ref).sum(dim=(1, 2)).mean()
+                aev_grad_grad = torch.autograd.grad(loss, aev_grad_, create_graph=True, retain_graph=True)[0]
```
Comment (zasdfgbnm) [inline tests/test_cuaev.py:259]: Same as above, make it a function and call it twice

File: tests/test_cuaev.py [hunk 4]
```diff
@@ -159,12 +308,53 @@ def testNISTBackward(self):
...
```
Comment (zasdfgbnm) [inline tests/test_cuaev.py:330]: Same as above, make it a function and call it twice

File: torchani/cuaev/aev.cu [hunk 5]
```diff
@@ -1,6 +1,7 @@
 #include <thrust/equal.h>
 #include <torch/extension.h>
 #include <cub/cub.cuh>
+#include <iostream>
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.cu:4]: Remove debugging code

File: torchani/cuaev/aev.cu [hunk 6]
```diff
@@ -205,6 +207,16 @@ __global__ void pairwiseDistance_backward(
...
+// template <typename DataT, typename IndexT = int>
+// __global__ void pairwiseForceDistance(
+//     torch::PackedTensorAccessor32<DataT, 3, torch::RestrictPtrTraits> grad_force,
+//     torch::PackedTensorAccessor32<DataT, 3, torch::RestrictPtrTraits> pos_t,
+//     PairDist<DataT>* d_Rij,
+//     PairDist<DataT>* d_radialFRij,
+//     IndexT nRadialRij) {
+
+// }
+
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.cu:219]: remove

File: torchani/cuaev/aev.cu [hunk 7]
```diff
@@ -1120,50 +1405,171 @@ Tensor cuaev_backward(
...
+  int* d_centerAtomStartIdx = (int*)tensor_centerAtomStartIdx.data_ptr();
+
+  // Tensor tensor_radialFRij = torch::empty(sizeof(PairDist<float>) * nRadialRij, d_options);
+  // PairDist<float>* d_radialFRij = (PairDist<float>*)tensor_radialFRij.data_ptr();
+
+  // dim3 block(8, 8, 1);
+  // pairwiseDistance<<<n_molecules, block, sizeof(float) * max_natoms_per_mol * 3, stream>>>(
+  //     grad_force.packed_accessor32<float, 3, torch::RestrictPtrTraits>(),
+  //     d_Rij, // d_radialFRij,
+  //     nRadialRij);
```
Comment (zasdfgbnm) [inline torchani/cuaev/aev.cu:1456]: remove

File: setup.py [hunk 8]
```diff
@@ -128,6 +128,7 @@ def cuaev_kwargs():
         'torch',
         'lark-parser',
         'requests',
+        'h5py',
```
Comment (zasdfgbnm) [inline setup.py:131]: Why h5py is a dependency of torchani, and why it is related to this PR?

File: tools/training-aev-benchmark.py [hunk 9]
```diff
@@ -85,7 +99,20 @@ def print_timer(label, t):
...
+        t = f'{t * 1000:.1f} ms'
+    else:
+        t = f'{t:.3f} sec'
+    return t
+
+
+def benchmark(parser, dataset, use_cuda_extension, force_train=False):
+    global summary
+    global runcounter
+
```
Comment (zasdfgbnm) [inline tools/training-aev-benchmark.py:113]: Why do you need these two lines?

Comment (yueyericardo) [reply tools/training-aev-benchmark.py:113]: they will be modified at the end of this function

File: tests/test_cuaev.py [hunk 10]
```diff
@@ -52,6 +52,50 @@ def setUp(self):
...
+        cuaev_grad = torch.autograd.grad(E, cu_aev_, create_graph=True, retain_graph=True)[0]
+        # graph1 backward
+        cuaev_grad_ = cuaev_grad.clone().detach().requires_grad_()
+        force_cuaev = torch.autograd.grad(cu_aev, coordinates, cuaev_grad_, create_graph=True, retain_graph=True)[0]
+        # force loss backward
+        # force_true = torch.randn_like(force_cuaev)
+        loss = torch.abs(force_true - force_cuaev).sum(dim=(1, 2)).mean()
+        cuaev_grad_grad = torch.autograd.grad(loss, cuaev_grad_, create_graph=True, retain_graph=True)[0]
+        cuaev_grad_grad_radial = cuaev_grad_grad[0][0][:self.radial_length].view(-1, self.aev_computer.radial_sublength)
+        cuaev_grad_grad_angular = cuaev_grad_grad[0][0][self.radial_length:].view(-1, self.aev_computer.angular_sublength)
```
Comment (zasdfgbnm) [inline tests/test_cuaev.py:92]: You should not copy paste the same code twice, make it a function.

File: tests/test_cuaev.py [hunk 11]
```diff
@@ -89,15 +133,82 @@ def testSimpleBackward(self):
...
-        cuaev_grad = coordinates.grad
+        force_cuaev = coordinates.grad
+        self.assertEqual(cu_aev, aev, f'cu_aev: {cu_aev}\n aev: {aev}')
+        self.assertEqual(force_cuaev, force_ref, f'\nforce_cuaev: {force_cuaev}\n aev_grad: {force_ref}')
+
+    def testSimpleDoubleBackward_1(self):
+        """
+        Test Double Backward (Force training) by parameters' gradient
+        """
+        coordinates = torch.tensor([
```
Comment (zasdfgbnm) [inline tests/test_cuaev.py:150]: Add a comment about what molecule it is

File: torchani/cuaev/README.md [hunk 12]
```diff
@@ -44,17 +44,29 @@ cuaev_computer = torchani.AEVComputer(Rcr, Rca, EtaR, ShfR, EtaA, Zeta, ShfA, Sh
 ## TODOs
 - [x] CUAEV Forward
 - [x] CUAEV Backwad (Force)
+- [x] CUAEV Double Backwad (Force training need aev's second derivative)
```
Comment (zasdfgbnm) [inline torchani/cuaev/README.md:47]:
```suggestion
- [x] CUAEV Double Backwad (Force training need aev's double backward w.r.t. grad_aev)
```

File: tools/aev-benchmark-size.py [hunk 13]
```diff
@@ -26,22 +26,41 @@ def info(text):
...
     aev = None
     for i in range(N):
-        aev = aev_comp(speciesPositions).aevs
+        species, coordinates = speciesPositions
+        if nn is not None:  # double backward
+            # graph1 input -> aev
+            coordinates = coordinates.requires_grad_()
+            _, aev = aev_comp((species, coordinates))
+            # graph2 aev -> E
+            aev_ = aev.clone().detach().requires_grad_()
```
Comment (zasdfgbnm) [inline tools/aev-benchmark-size.py:43]: Why do you need to detach in benchmark? Real force training does not detach, and benchmark should be consistent with real applications.

File: tools/training-aev-benchmark.py [hunk 14]
```diff
@@ -249,20 +285,43 @@ def wrapper(*args, **kwargs):
...
         checkgpu(i)
 
-    print("\n\n=> Test 1: USE cuda extension, Energy training")
+    # Warming UP
+    if len(dataset_shuffled) < 100:
+        runcounter = -1
+        parser.runname = 'Warning UP'
+        print(f"\n\n=> Test 0: {parser.runname}")
+        torch.cuda.empty_cache()
+        gc.collect()
```
Comment (zasdfgbnm) [inline tools/training-aev-benchmark.py:294]: If you are not lack of memory, please don't do this

File: setup.py (hunk 8 above)
Comment (yueyericardo) [reply setup.py]: removed

File: tests/test_cuaev.py [hunk 15]
```diff
@@ -52,6 +52,36 @@ def setUp(self):
         num_species = 4
         self.aev_computer = torchani.AEVComputer(Rcr, Rca, EtaR, ShfR, EtaA, Zeta, ShfA, ShfZ, num_species)
         self.cuaev_computer = torchani.AEVComputer(Rcr, Rca, EtaR, ShfR, EtaA, Zeta, ShfA, ShfZ, num_species, use_cuda_extension=True)
+        self.nn = torch.nn.Sequential(torch.nn.Linear(384, 1, False)).to(self.device)
+        self.radial_length = self.aev_computer.radial_length
+
+    def _double_backward(self, aev_computer, species, coordinates):
+        torch.manual_seed(12345)
+        # graph1 input -> aev
```
Comment (zasdfgbnm) [inline tests/test_cuaev.py:60]: Add a comment why you need to use two separate graph:
```
# We want to get the gradient of `grad_aev`, which requires `grad_aev` to be a leaf node
# due to `torch.autograd`'s limitation. So we split the coord->aev->energy graph into two separate
# graphs: coord->aev and aev->energy, so that aev and grad_aev are now leaves.
```

Comment (yueyericardo) [reply tests/test_cuaev.py:60]: good comment!

File: tests/test_cuaev.py [hunk 16]
```diff
@@ -52,6 +52,36 @@ def setUp(self):
...
+        coordinates = coordinates.clone().detach().requires_grad_()
+        _, aev = aev_computer((species, coordinates))
+        # graph2 aev -> E
+        aev_ = aev.clone().detach().requires_grad_()
+        E = self.nn(aev_).sum()
+        # graph2 backward
+        aev_grad = torch.autograd.grad(E, aev_, create_graph=True, retain_graph=True)[0]
+        # graph1 backward
+        aev_grad_ = aev_grad.clone().detach().requires_grad_()
+        force_ref = torch.autograd.grad(aev, coordinates, aev_grad_, create_graph=True, retain_graph=True)[0]
```
Comment (zasdfgbnm) [inline tests/test_cuaev.py:75]: Rename it to `force`, since this is a separate function, and it is not reference

File: tools/training-aev-benchmark.py (hunk 14 above)
Comment (yueyericardo) [reply tools/training-aev-benchmark.py:294]: this is for record relatively accurate gpu memory usage, you think this is not necessary?

File: tests/test_cuaev.py [hunk 17]
```diff
@@ -52,6 +52,64 @@ def setUp(self):
         num_species = 4
         self.aev_computer = torchani.AEVComputer(Rcr, Rca, EtaR, ShfR, EtaA, Zeta, ShfA, ShfZ, num_species)
         self.cuaev_computer = torchani.AEVComputer(Rcr, Rca, EtaR, ShfR, EtaA, Zeta, ShfA, ShfZ, num_species, use_cuda_extension=True)
+        self.nn = torch.nn.Sequential(torch.nn.Linear(384, 1, False)).to(self.device)
+        self.radial_length = self.aev_computer.radial_length
+
+    def _double_backward_1_test(self, species, coordinates):
+
+        def double_backward_1(aev_computer, species, coordinates):
```
Comment (zasdfgbnm) [inline tests/test_cuaev.py:60]: why not just name it as `double_backward`?

File: tests/test_cuaev.py [hunk 18]
```diff
@@ -52,6 +52,64 @@ def setUp(self):
...
+        aev, force_ref, param_grad_ref = double_backward_1(self.aev_computer, species, coordinates)
+        cu_aev, force_cuaev, param_grad = double_backward_1(self.cuaev_computer, species, coordinates)
+
+        self.assertEqual(cu_aev, aev, f'cu_aev: {cu_aev}\n aev: {aev}')
+        self.assertEqual(force_cuaev, force_ref, f'\nforce_cuaev: {force_cuaev}\n force_ref: {force_ref}')
+        self.assertEqual(param_grad, param_grad_ref, f'\param_grad: {param_grad}\n param_grad_ref: {param_grad_ref}', atol=5e-5, rtol=5e-5)
+
+    def _double_backward_2_test(self, species, coordinates):
+
+        def double_backward_2(aev_computer, species, coordinates):
```
Comment (zasdfgbnm) [inline tests/test_cuaev.py:82]: same here

File: tools/training-aev-benchmark.py (hunk 14 above)
Comment (zasdfgbnm) [reply tools/training-aev-benchmark.py:294]: If you record accurate gpu memory usage, you might not be able to record accurate speed.

**Review by zasdfgbnm — APPROVED** [review]
